src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/INetworkManager.cs

```diff
@@ -54,7 +59,7 @@ public interface INetworkManager : ICommandRegistry
         /// <summary>
         /// Send a packet of data to all connected devices.
         /// </summary>
-        /// <param name="data">The data to send to each connected device.</param>
-        void Broadcast(byte[] data);
+        /// <param name="data">The data to send to each connected device. The reference is set to null once taking over ownership</param>
+        void Broadcast(ref byte[] data);

```

Comment:
This reads strangely - it basically forces the caller to make a copy.
A familiar idiom which avoids the copy is:
```
void Broadcast(byte[] data, int offset=0, int len=-1); // if len==-1 data.len is used instead
```
Then, callers would look like this
```
using (MemoryStream stream = new MemoryStream())
using (BinaryWriter message = new BinaryWriter(memoryStream)) {
                message.Write(...);
                connectionManager.Broadcast(stream.GetBuffer(), 0, stream.Position);
}

Reply:
I looked a bit more and almost all of the callers of Broadcast are creating their own local stream & binarywriter, writing a few bytes and discarding them both. Another alternative is to put the stream & writer in the manager like this:
```
void Broadcast( Action<BinaryWriter> action );
```
Here's a sample of some of the current code (SetPerformanceMonitoringMode)
```
                byte[] message;
                using (MemoryStream stream = new MemoryStream())
                using (BinaryWriter writer = new BinaryWriter(stream))
                {
                    writer.Write(PerfDiagnosticModeEnabledCommand);
                    writer.Write(enabled);
                    writer.Flush();
                    message = stream.ToArray();
                }
                connectionManager.Broadcast(message);
```
With the action idiom, it it more concise, no copying & no garbage created.
```
                connectionManager.Broadcast( (writer) => {
                    writer.Write(PerfDiagnosticModeEnabledCommand);
                    writer.Write(enabled);
                    writer.Flush();
                });
```

Reply:
On second thoughts, I'm not sure if creation of that lambda produces garbage. Any C# experts know?

Reply:
As I understand lambda causes an allocation if it captures a local variable. 

When it contains a set of self-contained statements, it gets compiled to a static method at compile time that is referenced through a Delegate. 

If it references the values of current instance class, then, it gets compiled to the instance method of the class and passed through a delegate.

If it captures a local variable, then a class is created with a member for that variable and the instance method. That variable is assigned, and the instance method is passed through a Delegate.

Reply:
Your example will result in a static method at compile time, and no allocation.

Reply:
Thanks Andrei! I read a bit more and it looks like there are 2 parts to this.
1. the compiler may create an object for the captures or use a static or class method.
2. the compiler creates an instance of Action<T>
So while we're can avoid \#1, exactly as you say, it seems there's no way to avoid \#2. I verified this with ildasm :/
Still, there are fewer allocations and virtual calls (e.g. we save 2 Disposes)

Reply:
I will address this in a follow up review